tracer/src/instruction/test.rs

```diff
@@ -1,89 +1,246 @@
+#[cfg(test)]
+macro_rules! impl_from_operands_r {
+    ($($ty:ident),* $(,)?) => {
+        $(
+            #[cfg(test)]
+            impl $ty {
+                pub fn from_operands(rd: usize, rs1: usize, rs2: usize, address: u64) -> Self {
+                    Self {
+                        address,
+                        operands: FormatR { rd, rs1, rs2 },
+                        virtual_sequence_remaining: None,
+                    }
+                }
+            }
+        )*
+    };
+}
+
+impl_from_operands_s!(SB, SH);
+impl_from_operands_load!(LB, LBU, LH, LHU);
+impl_from_operands_r!(DIV, DIVU, REM, REMU);
+
+fn test_rng() -> StdRng {
+    // StdRng::from_seed([10u8; 32]) // deterministic
+    StdRng::from_entropy()
+}
+
+/// Returns *(rs1_val, imm)* such that `rs1_val + imm` is even and ≥ 0.
+fn generate_aligned_address(rng: &mut StdRng) -> (u64, i64, u64) {
+    const IMM_MAX: i64 = (1 << 11) - 1;
+    const IMM_MIN: i64 = -(1 << 11);
+
+    // reroll until satisfied
+    loop {
+        // Pick a random even offset inside RAM
+        let offset = (rng.next_u64() % (DRAM_SIZE as u64 - 2)) & !1;
+        let dest_addr = DRAM_BASE + offset;
+
+        // Choose an immediate in range, then back‑solve rs1.
+        let imm = (rng.next_u64() as i64 % (IMM_MAX - IMM_MIN + 1)) + IMM_MIN;
+        let rs1_val = dest_addr as i64 - imm;
+
+        if rs1_val >= 0 {
+            return (rs1_val as u64, imm, dest_addr);
+        }
+    }
+}
+
+/// Fuzzer for both mem and nonmem instructions
+fn virtual_sequence_fuzz<
+    I: RISCVInstruction + VirtualInstructionSequence + RISCVTrace + Copy + 'static,
+>(
+    is_memory: bool,
+) where
+    RV32IMCycle: From<RISCVCycle<I>>,
+    <I as RISCVInstruction>::RAMAccess: Default,
+{
+    let mut rng = test_rng();
+    const REG_COUNT: usize = 32;
+
+    // Init CPUs
+    let mut cpu_exec = Cpu::new(Box::new(DummyTerminal::new()));
+    let mut cpu_trace = Cpu::new(Box::new(DummyTerminal::new()));
+    cpu_exec.update_xlen(Xlen::Bit32);
+    cpu_trace.update_xlen(Xlen::Bit32);
+    cpu_exec.mmu.init_memory(DRAM_BASE + DRAM_SIZE);
+    cpu_trace.mmu.init_memory(DRAM_BASE + DRAM_SIZE);
+
+    for _ in 0..10000 {
+        // Seed operands accordingly
+        let rs1_idx = rng.next_u64() % REG_COUNT as u64;
+        let rs2_or_rd_idx = rng.next_u64() % REG_COUNT as u64;
+        let addr_tag = rng.next_u64();
+
+        let (rs1_val, imm, _dest) = if is_memory {
+            generate_aligned_address(&mut rng)
+        } else {
+            (rng.next_u32() as u64, (rng.next_u32() & 0x0fff) as i64, 0)
+        };
+
+        // Construct instruction
+        // Since we need control over operands we use a custom `from_operands` method
+        let instruction: I = if is_memory {
+            let ty = TypeId::of::<I>();
+            if ty == TypeId::of::<SB>() || ty == TypeId::of::<SH>() {
+                // Store → FormatS downcast
+                unsafe {
```

Comment:
Maybe I'm wrong, but this downcasting should be removed as well if you use `random()` on the instructoin and state.